### PR TITLE
Refactor device plugin to register with kubelet and dynamically start client's gRCP servers

### PR DESCRIPTION
**Pull request by sbezverk**
This PR refactors the current device plugin implementation:
1. Adds a pool of sockets
2. Advertises sockets to kubelet
3. On Allocate request, starts gRPC server for a client on a client visible socket.
4. Checks for re-use of a scoket request from kubelet and gracefully deals with it.
 
Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

### CODE REVIEW

File: plugins/netmesh/plugin_impl_netmesh.go [hunk 1]
```diff
@@ -81,17 +78,13 @@ func (plugin *Plugin) Init() error {
 // the kvdbsync is fully initialized and ready for publishing when a k8s
 // notification comes.
 func (plugin *Plugin) AfterInit() error {
-	netmeshdp = nsmdp.NewNSMDevicePlugin()
-	netmeshdp.Serve()
+	err := nsmdp.NewNSMDevicePlugin()
```
Comment (mestery) [inline plugins/netmesh/plugin_impl_netmesh.go]: What if we passed in the Plugin() here so we could make use of the logger from the Plugin() in the nsmdp code? That would give us consistency around logging. Thoughts?

Comment (sbezverk) [reply plugins/netmesh/plugin_impl_netmesh.go]: I thought about it once I added refactored code, I will look into it, thanks for suggestions.

**Comment by mestery** [comment]
If the build failure is timing related, the timeout needs to be increased. Otherwise, the pod isn't coming up, and you'll want to look at the logs to see why. May need to add some debug to dump those.

**Comment by sbezverk** [comment]
@mestery I do not think it is time related but on the other hand daemonset runs fine in my local cluster.  I will add extra debugging to see why daemonset is not coming up.

File: plugins/netmesh/plugin_impl_netmesh.go (hunk 1 above)
Comment (mestery) [reply plugins/netmesh/plugin_impl_netmesh.go]: Great, thanks! We may need to come up with a broader logging strategy, so if it becomes too challenging, feel free to open an issue and we'll tackle it there.

**Comment by mestery** [comment]
Yay, it passes CI! I'll review in detail in a few hours. Thanks @sbezverk !

File: scripts/integration-tests [hunk 2]
```diff
@@ -28,6 +28,21 @@ KUBECTL_VERSION=v1.10.3
 curl -Lo kubectl https://storage.googleapis.com/kubernetes-release/release/${KUBECTL_VERSION}/bin/linux/amd64/kubectl && \
 	chmod +x kubectl && sudo mv kubectl /usr/local/bin/
 
+#
+# In case if a failure, collecting some evidence for further debugging
+#
+function error_collection {
```
Comment (mestery) [inline scripts/integration-tests]: Love this! One small nit if you re-spin: Maybe put some sort of flashy logging at the start, something like:

```
echo "=====> ERROR: Timed out waiting for NSM daemonset to start"
```

File: nsmdp/nsmdp.go [hunk 3]
```diff
@@ -20,89 +20,123 @@
...
 	}
-	return n
-}
 
-func generateToken() string {
-	b := make([]byte, 16)
-	rand.Read(b)
-	return fmt.Sprintf("%x", b)
-}
+	unix.Umask(0077)
```
Comment (mestery) [inline nsmdp/nsmdp.go]: Nit: Make this mask a constant.

File: nsmdp/nsmdp.go [hunk 4]
```diff
@@ -20,89 +20,123 @@
...
+	sock, err := net.Listen("unix", listenEndpoint)
+	if err != nil {
+		logger.Infof(" failure to listen on socket %s with error: %+v", client.socketPath, err)
```
Comment (mestery) [inline nsmdp/nsmdp.go]: Nit: Space before failure should be removed. I see this in a few logs below. And I think we should be consistent and capitalize the first word in a log message as well.

File: nsmdp/nsmdp.go [hunk 5]
```diff
@@ -20,89 +20,123 @@
...
-			_, ok := n.devs[id]
-			if !ok {
-				return nil, fmt.Errorf("invalid allocation request: unknown device: %s", id)
-			}
-			mount := &pluginapi.Mount{
-				ContainerPath: SocketBaseDir + id,
-				HostPath:      SocketBaseDir + id,
+			if _, ok := n.nsmSockets[id]; ok {
+				if n.nsmSockets[id].allocated {
+					// Socket has been previsously used, since we did noy get notification from
```
Comment (mestery) [inline nsmdp/nsmdp.go]: Nit: s/noy/not/

File: nsmdp/nsmdp.go [hunk 6]
```diff
@@ -20,89 +20,123 @@
...
+					// kubelet when POD using this socket went down, gRPC client's server
+					// needs to be stopped.
```
Comment (mestery) [inline nsmdp/nsmdp.go]: This is a good catch, thanks for catching and handling this case where the soccer is re-used!

File: scripts/integration-tests (hunk 2 above)
Comment (sbezverk) [reply scripts/integration-tests]: done

File: nsmdp/nsmdp.go (hunk 3 above)
Comment (sbezverk) [reply nsmdp/nsmdp.go]: done

**Comment by sbezverk** [comment]
@fkautz please review.

File: nsmdp/nsmdp.go [hunk 7]
```diff
@@ -20,89 +20,123 @@
...
+					HostPath:      path.Join(socketBaseDir, fmt.Sprintf("nsm-%s", id)),
+					ReadOnly:      false,
+				}
+				n.nsmSockets[id] = nsmSocket{
+					device:      &pluginapi.Device{ID: id, Health: pluginapi.Healthy},
+					socketPath:  path.Join(mount.HostPath, serverSock),
+					stopChannel: make(chan bool),
+					allocated:   true,
+				}
+				os.MkdirAll(mount.HostPath, folderMask)
```
Comment (fkautz) [inline nsmdp/nsmdp.go]: We should catch and report any errors from os.MkdirAll

File: nsmdp/nsmdp.go [hunk 8]
```diff
@@ -20,89 +20,123 @@
...
+	select {
+	case <-client.stopChannel:
+		logger.Infof("Server for socket %s received shutdown request", client.socketPath)
+	}
+	client.allocated = false
+	client.stopChannel <- true
 }
 
-func (n *NSMDevicePlugin) Allocate(ctx context.Context, reqs *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+func (n *nsmClientEndpoints) Allocate(ctx context.Context, reqs *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
```
Comment (fkautz) [inline nsmdp/nsmdp.go]: Do we know if Allocate can be called concurrently with another Allocate or if ListAndWatch can read or update state concurrently while Allocate is running?

We should consider synchronizing this code and the hot spot in ListAndWatch.

File: nsmdp/nsmdp.go [hunk 9]
```diff
@@ -112,15 +146,111 @@ func (n *NSMDevicePlugin) Allocate(ctx context.Context, reqs *pluginapi.Allocate
...
+	}
+	return c, nil
+}
+
+func startDeviceServer(nsm *nsmClientEndpoints) error {
+	// Initial socket clean up
+	listenEndpoint := path.Join(pluginapi.DevicePluginPath, serverSock)
+	fi, err := os.Stat(listenEndpoint)
+	if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
+		os.Remove(listenEndpoint)
```
Comment (fkautz) [inline nsmdp/nsmdp.go]: Do we want to drop the error in os.Remove?

File: nsmdp/nsmdp.go (hunk 7 above)
Comment (sbezverk) [reply nsmdp/nsmdp.go]: ok, will do.

File: nsmdp/nsmdp.go (hunk 8 above)
Comment (sbezverk) [reply nsmdp/nsmdp.go]: Allocate call is not concurrent. ListAndWatch for out case is not really required, I mean running in the continues loop, as we do not expect to see any changes in the number or state of the pool os socket. In theory I could reply just once, and then do a fake loop just to keep grpc stream connection alive.

File: nsmdp/nsmdp.go (hunk 9 above)
Comment (sbezverk) [reply nsmdp/nsmdp.go]: I can add error handling here.

**Pull request merged**